Rebus.LegacyExtensions/LegacyCompatibility/MapLegacyHeadersOutgoingStep.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Rebus.Messages;
+using Rebus.Pipeline;
+
+namespace Rebus.LegacyCompatibility
+{
+	class MapLegacyHeadersOutgoingStep : IOutgoingStep
+	{
+		private bool _propagateAutoCorrelationSagaId;
+		internal const string AutoCorrelationSagaIdHeader = "rebus-autocorrelation-saga-id";
+
+		public MapLegacyHeadersOutgoingStep(bool propagateAutoCorrelationSagaId)
+		{
+			_propagateAutoCorrelationSagaId = propagateAutoCorrelationSagaId;
+		}
+		private void AddCustomHeaders(Dictionary<string, string> headers)
+		{
+			headers["message-date"] = DateTime.Parse(headers["rbs2-senttime"]).ToUniversalTime().ToString("O");
+		}
+
+		private void AddMulticastHeader (Dictionary<string, string> headers)
+		{
+			if (headers[Headers.Intent] == "pub")
+			{
+				headers["rebus-multicast"] = "";
+			}
+		}
+		private void PropagateAutoCorrelationSagaIdToHeaders(Dictionary<string, string> headers)
+		{
+			if (MessageContext.Current?.TransactionContext?.Items?.ContainsKey(AutoCorrelationSagaIdHeader) == true)
+				headers[AutoCorrelationSagaIdHeader] = (string)MessageContext.Current.TransactionContext.Items[AutoCorrelationSagaIdHeader];
+		}
+
+		public async Task Process(OutgoingStepContext context, Func<Task> next)
+		{
+			var transportMessage = context.Load<TransportMessage>();
+
+			var body = transportMessage.Body;
+			var headers = transportMessage.Headers;
+
+			var newHeaders = MapTrivialHeaders(headers);
+
+		   var allHeaders =  newHeaders.Union(headers).ToDictionary(x => x.Key, x => x.Value);
+
+			MapSpecialHeaders(allHeaders);
+
+			AddCustomHeaders(allHeaders);
+			AddMulticastHeader(allHeaders);
+
+			if (_propagateAutoCorrelationSagaId)
+				PropagateAutoCorrelationSagaIdToHeaders(allHeaders);
+
+			context.Save(new TransportMessage(allHeaders, body));
+
+			await next();
+		}
+
+		void MapSpecialHeaders(Dictionary<string, string> headers)
+		{
+			MapContentType(headers);
+		}
+
+		void MapContentType(Dictionary<string, string> headers)
+		{
+			string contentType;
+			if (!headers.TryGetValue(Headers.ContentType, out contentType)) return;
+
+			if (!contentType.StartsWith(CustomJsonSerializer.JsonContentType))
+			{
+				throw new FormatException($"Unknown content type: '{contentType}'");
+			}
+
+			var encoding = contentType
```

Comment:
This is assuming an specific CT format. Also this logic is already present somewhere else within this same patch.. do we need to make this same parsing twice?